usl_models/usl_models/flood_ml/model.py

```diff
@@ -23,65 +30,111 @@ class FloodConvLSTM(tf.keras.Model):
         )
 
         # Geospatial CNN
         geo_cnn_params = {"strides": 2, "padding": "same", "activation": "relu"}
         self.geo_cnn = tf.keras.Sequential(
             [
+                # Input shape: (height, width, channels)
+                layers.InputLayer(
+                    (self.spatial_height, self.spatial_width, constants.GEO_FEATURES)
+                ),
                 layers.Conv2D(16, 5, **geo_cnn_params),
                 layers.MaxPool2D(pool_size=2, strides=1, padding="same"),
                 layers.Conv2D(64, 5, **geo_cnn_params),
                 layers.MaxPool2D(pool_size=2, strides=1, padding="same"),
-            ]
+            ],
+            name="geospatial_cnn",
         )
+        # This is a constant (time-invariant), so we cache the output.
+        self.geo_cnn_output = None
 
         # ConvLSTM
-        self.concat = layers.Concatenate(axis=-1)
-        self.conv_lstm = layers.ConvLSTM2D(
-            self.params["lstm_units"],
-            self.params["lstm_kernels"],
-            strides=1,
-            padding="same",
-            activation="tanh",
-            dropout=self.params["lstm_dropout"],
-            recurrent_dropout=self.params["lstm_recurrent_dropout"],
+        # The spatial dimensions have been reduced 4x by the CNNs.
+        # The "channel" dimension is the sum of the channels from the CNNs
+        # and the rainfall window size.
+        conv_lstm_height = self.spatial_height // 4
+        conv_lstm_width = self.spatial_width // 4
+        conv_lstm_channels = 16 + 64 + self.params["m_rainfall"]
+        self.conv_lstm = tf.keras.Sequential(
+            [
+                # Input shape: (time_steps, height, width, channels)
+                layers.InputLayer(
+                    (None, conv_lstm_height, conv_lstm_width, conv_lstm_channels)
+                ),
+                layers.ConvLSTM2D(
+                    self.params["lstm_units"],
+                    self.params["lstm_kernels"],
+                    strides=1,
+                    padding="same",
+                    activation="tanh",
+                    dropout=self.params["lstm_dropout"],
+                    recurrent_dropout=self.params["lstm_recurrent_dropout"],
+                ),
+            ],
+            name="conv_lstm",
         )
 
-        # Output (upsampling) CNN
+        # Output CNN (upsampling via TransposeConv)
         output_cnn_params = {"padding": "same", "activation": "relu"}
         self.output_cnn = tf.keras.Sequential(
             [
+                # Input shape: (height, width, channels)
+                layers.InputLayer(
+                    (conv_lstm_height, conv_lstm_width, self.params["lstm_units"])
+                ),
                 layers.Conv2DTranspose(8, 4, strides=4, **output_cnn_params),
                 layers.Conv2DTranspose(1, 1, strides=1, **output_cnn_params),
-            ]
+            ],
+            name="output_cnn",
         )
 
     def forward(
```

Comment:
we are considering all the data, so the next step will be to include validation and test?

Reply:
Everything will be done incrementally. This PR just finishes implementing the model logic so that it's actually callable. The actual pipeline work (training, evaluation) will come later, likely also via separate PRs.

Reply:
Thank you!